### DISCUSSION TITLE
How can we use ModelIntermediateOutput

### DISCUSSION THREAD

**Discussion by Amrusama** [privacytrustlab/ml_privacy_meter#102] [Q&A]
Hello,

I am trying to use the ModelIntermediateOutput to perform membership attacks on specific layers and it's not clear to me how can i pass the parameter extra to ModelIntermediateOutput.

**Comment by changhongyan123** [comment]
Hi @Amrusama,

Currently, we do not have optimal support for using the model intermediate output to perform attacks. However, with some slight modifications to the code, you can achieve your objectives. Here is one possible workaround.

First, you can check out the tutorial on how to use the gradient norm to perform attacks [here](https://github.com/privacytrustlab/ml_privacy_meter/blob/0e928110d4ba3e44dae022533867c7803583ee24/tutorials/white_box_attack.ipynb), which has a similar logic to the attack you want to perform.

In terms of how to obtain the Model Intermediate Output, here is a brief introduction to the logic, and you can follow the necessary steps to achieve your goal:

In the [ModelIntermediateOutput class](https://github.com/privacytrustlab/ml_privacy_meter/blob/0e928110d4ba3e44dae022533867c7803583ee24/privacy_meter/information_source_signal.py#L156), the intermediate output is computed by calling `model.get_intermediate_outputs(extra["layers"])`. Therefore, you can pass the layer number to the extra dictionary when specifying the signal you want to use. It will call the `get_intermediate_outputs` function defined in the model wrapper (e.g., if you are using PyTorch, it is implemented [here](https://github.com/privacytrustlab/ml_privacy_meter/blob/0e928110d4ba3e44dae022533867c7803583ee24/privacy_meter/model.py#L659)). Note that if you are using a framework other than PyTorch, you may need to implement a way of obtaining the model intermediate output in the corresponding model wrapper. You can refer to the implementation in PyTorch as a guide and modify it to fit your chosen framework (see [here](https://github.com/privacytrustlab/ml_privacy_meter/blob/0e928110d4ba3e44dae022533867c7803583ee24/privacy_meter/model.py#L659)).

When computing the signal, you would need to modify the code by inputting the `extra` dictionary [here](https://github.com/privacytrustlab/ml_privacy_meter/blob/0e928110d4ba3e44dae022533867c7803583ee24/privacy_meter/metric.py#L101).

At the same time, we are also working on improving the support for using the model intermediate output to perform attacks.
